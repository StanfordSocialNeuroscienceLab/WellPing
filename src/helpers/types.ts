import { QuestionType } from "./helpers";

export type StreamName = string;

export type QuestionId = string;

export interface Question {
  id: QuestionId;
  type: QuestionType;
  question: string;
  next?: QuestionId | null;
}

export interface SliderQuestion extends Question {
  type: QuestionType.Slider;
  slider: [string, string]; // [left, right]
  defaultValue?: number;
  defaultValueFromQuestionId?: QuestionId;
}

export type Choice = {
  key: string;
  value: string;
};

export interface ChoicesQuestion extends Question {
  choices: Choice[];
  specialCasesStartId?: {
    [questionId: string /* actually QuestionId */]: QuestionId;
    _pna?: QuestionId;
  };
  randomizeChoicesOrder?: boolean;
  randomizeExceptForChoiceIds?: string[];
}

export interface ChoicesWithSingleAnswerQuestion extends ChoicesQuestion {
  type: QuestionType.ChoicesWithSingleAnswer;
}

export interface ChoicesWithMultipleAnswersQuestion extends ChoicesQuestion {
  type: QuestionType.ChoicesWithMultipleAnswers;
}

export interface YesNoQuestion extends Question {
  type: QuestionType.YesNo;
  branchStartId?: {
    yes?: QuestionId;
    no?: QuestionId;
  };
  // Currently only `yes` is supported and also can only followup after 3 days and 7 days with the same stream.
  addFollowupStream?: {
    yes?: StreamName;
    // TODO: no?: StreamName;
  };
}

export interface MultipleTextQuestion extends Question {
  // `id` will store the number of text fields answered.
  type: QuestionType.MultipleText;
  eachId: QuestionId;
  placeholder?: string;
  choices?: "NAMES" | Choice[];
  forceChoice?: boolean;
  max: number;
  maxMinus?: QuestionId; // The max number of text field will be `max` minus the number of text the participant entered in `maxMinus` question.
  repeatedItemStartId?: QuestionId;
  fallbackItemStartId?: QuestionId; // This is used when the user does not enter any name or select prefer not to answer. Note that this has to exists somewhere else. If it is `null`, we will go to `next` directly.
  indexName: string;
  variableName: string;
}

export interface HowLongAgoQuestion extends Question {
  type: QuestionType.HowLongAgo;
}

export interface BranchQuestion extends Question {
  // This is not actually a question (it will not be displayed to the user)
  type: QuestionType.Branch;
  condition: {
    questionId: QuestionId;
    questionType:
      | QuestionType.MultipleText
      | QuestionType.ChoicesWithSingleAnswer;
    compare: "equal";
    target: number | string;
  };
  branchStartId: {
    true?: QuestionId;
    false?: QuestionId;
  };
}

export interface BranchWithRelativeComparisonQuestion extends Question {
  // This is not actually a question (it will not be displayed to the user)
  type: QuestionType.BranchWithRelativeComparison;
  branchStartId: {
    [questionId: string /* actually QuestionId */]: QuestionId;
  };
}

export interface QuestionsList {
  [id: string]: Question;
}

export type Streams = {
  [stream: string /* actually StreamName */]: QuestionsList;
};

export type StreamStartingQuestionIds = {
  [stream: string /* actually StreamName */]: QuestionId;
};

export type StudyID = string;

export type StudyInfo = {
  id: StudyID;
  consentFormUrl: string;
  contactEmail?: string;
  weekStartsOn: 0 | 1 | 2 | 3 | 4 | 5 | 6; // 0 is Sunday
  startDate: Date; // First survey will be sent after this time.
  endDate: Date; // Last survey will be sent before this time.
  frequency: {
    hoursEveryday: number[];

    // Randomly add a number of minutes between `min` to `max` minutes (inclusive) to the notification time.
    randomMinuteAddition: {
      min: number;
      max: number;
    };
  };
  streamsOrder: {
    // Notice the numbering here is not affected by the `weekStartsOn` option above.
    // The number of elements each day should be equal to the number of elements in `frequency.hoursEveryday` above.
    0: StreamName[]; // Sunday
    1: StreamName[]; // Monday
    2: StreamName[]; // Tuesday
    3: StreamName[]; // Wednesday
    4: StreamName[]; // Thursday
    5: StreamName[]; // Friday
    6: StreamName[]; // Saturday
  };

  // A stream that the user will see by default if there is any error in `streamsOrder`.
  streamInCaseOfError: StreamName;

  // A list of streams that will not be replaced by followup streams
  // (when a YesNo question's `addFollowupStream` is set).
  streamsNotReplacedByFollowupStream?: StreamName[];

  notificationContent: {
    default: {
      title: string;
      body: string;
    };
    // We will use `bonus` content for the notification until the number of survey completed that week is greater than `bonus.numberOfCompletionEachWeek - 1`.
    // `#n_ping#` will turn into "5 pings" or "1 ping".
    bonus: {
      title: string;
      body: string;
      numberOfCompletionEachWeek: number;
    };
  };
  serverURL: string;
};

export type SurveyFile = {
  studyInfo: StudyInfo;
  meta: {
    startingQuestionIds: StreamStartingQuestionIds;
  };
  streams: Streams;
};

export type Names = string[];
