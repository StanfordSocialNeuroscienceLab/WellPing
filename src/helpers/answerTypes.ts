import { AnswerEntity } from "../entities/AnswerEntity";
import { Question, QuestionsList, Choice } from "./types";

// We use `{ value: ... }` because it seems that `"simple-json"`
// does not support simply mixing normal values.
// This also increases the consistency of the data.

export type SliderAnswerData = {
  value: number;
};

export type ChoicesWithSingleAnswerAnswerData = {
  value: Choice;
};

// We use an array of tuples here (instead of `{ [key: string]: boolean }`)
// because the order shown on the screen can be kept in the results.
// This is potentially helpful for analysis later if the order is randomized
// as now we know the order the user saw the choices in.
export type ChoicesWithMultipleAnswersAnswerChoices = [Choice, boolean][];
export type ChoicesWithMultipleAnswersAnswerData = {
  value: ChoicesWithMultipleAnswersAnswerChoices;
};

export type YesNoAnswerData = {
  value: boolean;
};

export type MultipleTextAnswerData = {
  value: string[];
};

export type HowLongAgoAnswerDataType = [number | null, string | null];
export type HowLongAgoAnswerData = {
  value: HowLongAgoAnswerDataType;
};

export type AnswerData =
  | SliderAnswerData
  | ChoicesWithSingleAnswerAnswerData
  | ChoicesWithMultipleAnswersAnswerData
  | YesNoAnswerData
  | MultipleTextAnswerData
  | HowLongAgoAnswerData;

export interface AnswersList {
  [id: string]: AnswerEntity;
}

export interface QuestionScreenProps {
  question: Question;
  onDataChange: (data: AnswerData) => void;
  allAnswers: AnswersList;
  allQuestions: QuestionsList;
  pipeInExtraMetaData: (input: string) => string;
  setDataValidationFunction: (func: () => boolean) => void;
}
