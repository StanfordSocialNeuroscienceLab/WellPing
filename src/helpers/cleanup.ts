import { secureRemoveAllAnswersAsync } from "./secureStore/answer";
import { secureRemoveAllPingsAsync } from "./secureStore/ping";

export async function clearAllPingsAndAnswersAsync(): Promise<void> {
  await secureRemoveAllAnswersAsync();
  await secureRemoveAllPingsAsync();
}
