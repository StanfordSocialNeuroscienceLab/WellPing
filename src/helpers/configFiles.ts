import { isThisWeek } from "date-fns";

import { SurveyFile, Names, StudyID, StudyInfo, StreamName } from "./types";

export function getSurveyFile(): SurveyFile {
  const survey: SurveyFile = require("../../config/survey.json");
  return survey;
}

// TODO: DECOUPLE FUNCTIONS LIKE THIS
export function getNamesFile(): Names {
  const names: Names = require("../../config/names.json");
  return names;
}

export function getStudyInfo(): StudyInfo {
  return getSurveyFile().studyInfo;
}

export function getAllStreamNames(): StreamName[] {
  return Object.keys(getSurveyFile().meta.startingQuestionIds) as StreamName[];
}

export function getStudyId(): StudyID {
  return getStudyInfo().id;
}

export function isTimeThisWeek(time: Date): boolean {
  return isThisWeek(time, { weekStartsOn: getStudyInfo().weekStartsOn });
}
