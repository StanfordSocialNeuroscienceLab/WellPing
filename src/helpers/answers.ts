import { AnswerData, Answer } from "./answerTypes";
import { getAnswersPingIdsQuestionIdsListAsync } from "./asyncStorage/answersPingIdsQuestionIdsList";
import { AnswerSchema } from "./schemas/Answer";
import {
  secureGetAnswerAsync,
  secureStoreAnswerAsync,
} from "./secureStore/answer";
import { Question, Ping } from "./types";

export async function getAnswersAsync(): Promise<Answer[]> {
  const answersList = await getAnswersPingIdsQuestionIdsListAsync();

  // https://stackoverflow.com/q/28066429/2603230
  const answers: Answer[] = await Promise.all(
    answersList.map(async (pingIdAndQuestionId) => {
      const answer = await secureGetAnswerAsync(
        pingIdAndQuestionId[0],
        pingIdAndQuestionId[1],
      );
      if (answer === null) {
        throw new Error(
          "answer === null in answersList.map in getAnswersAsync.",
        );
      }
      return answer;
    }),
  );

  return answers;
}

export async function insertAnswerAsync({
  ping,
  question,
  realQuestionId,
  preferNotToAnswer,
  data,
  date,
}: {
  ping: Ping;
  question: Question;
  realQuestionId: string;
  preferNotToAnswer: true | null; // See `MARK: WHY_PNA_TRUE_OR_NULL`.
  data: AnswerData | null;
  date: Date;
}): Promise<Answer> {
  const answer = AnswerSchema.parse({
    pingId: ping.id,
    questionId: realQuestionId,
    preferNotToAnswer,
    data,
    date,
  });

  await secureStoreAnswerAsync(answer);

  return answer;
}
