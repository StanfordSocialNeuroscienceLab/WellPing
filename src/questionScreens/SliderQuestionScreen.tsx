import React from "react";
import { View, Text, Slider } from "react-native";

import {
  QuestionScreenProps,
  SliderAnswerData,
  SliderAnswer,
} from "../helpers/answerTypes";
import {
  getNonCriticalProblemTextForUser,
  alertWithShareButtonContainingDebugInfo,
} from "../helpers/debug";
import { SliderQuestion } from "../helpers/types";

export const DEFAULT_SLIDER_VALUE = 50;

export const getQuestionDefaultSliderValue = (question: SliderQuestion) => {
  if (question.defaultValue == null) {
    return DEFAULT_SLIDER_VALUE;
  }
  return question.defaultValue;
};

interface SliderQuestionScreenProps extends QuestionScreenProps {
  question: SliderQuestion;
}

const SliderQuestionScreen: React.ElementType<SliderQuestionScreenProps> = ({
  question,
  loadingCompleted,
  onDataChange,
  allAnswers,
  allQuestions,
  pipeInExtraMetaData,
}) => {
  React.useEffect(() => {
    loadingCompleted();
  }, []);

  let defaultSliderValue = getQuestionDefaultSliderValue(question);
  if (question.defaultValueFromQuestionId) {
    const prevQuestionAnswer = allAnswers[
      pipeInExtraMetaData(question.defaultValueFromQuestionId)
    ] as SliderAnswer;
    if (prevQuestionAnswer && prevQuestionAnswer.data != null) {
      defaultSliderValue = prevQuestionAnswer.data.value;
    } else {
      // If the user did not answer the previous question,
      // use the default slider value from that question instead.
      const prevQuestion = allQuestions[
        question.defaultValueFromQuestionId
      ] as SliderQuestion;
      if (prevQuestion == null) {
        alertWithShareButtonContainingDebugInfo(
          getNonCriticalProblemTextForUser(
            `defaultValueFromQuestionId ${question.defaultValueFromQuestionId} prevQuestion == null`,
          ),
        );
      } else {
        defaultSliderValue = getQuestionDefaultSliderValue(prevQuestion);
      }
    }
  }

  return (
    <View style={{ paddingVertical: 30 }}>
      <Slider
        step={1}
        value={defaultSliderValue}
        minimumValue={0}
        maximumValue={100}
        minimumTrackTintColor="#2F2424"
        maximumTrackTintColor="#2F2424"
        onSlidingComplete={(value) => {
          onDataChange({ value } as SliderAnswerData);
        }}
        accessibilityLabel="slider input"
      />
      <View style={{ flexDirection: "row", justifyContent: "space-between" }}>
        <Text style={{ maxWidth: "40%" }}>{question.slider[0]}</Text>
        <Text style={{ maxWidth: "40%" }}>{question.slider[1]}</Text>
      </View>
    </View>
  );
};

export default SliderQuestionScreen;
