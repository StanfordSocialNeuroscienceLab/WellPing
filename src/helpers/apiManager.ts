import Constants from "expo-constants";
import * as Crypto from "expo-crypto";
import * as Device from "expo-device";
import { Platform } from "react-native";

import { AnswerEntity } from "../entities/AnswerEntity";
import { PingEntity } from "../entities/PingEntity";
import { getPingsAsync, getAnswersAsync } from "./asyncStorage";
import { getSurveyFileAsync } from "./configFiles";
import { getUserAsync, User, storeUserAsync } from "./user";

export async function getServerUrlAsync(): Promise<string> {
  return (await getSurveyFileAsync()).studyInfo.serverURL;
}

type UploadData = {
  patientId: string;
  pings: PingEntity[];
  answers: AnswerEntity[];
};

// If success, return `null`. Else return error message.
export async function registerUserAsync(user: User): Promise<string | null> {
  if (!user.patientId || !user.password) {
    return "You should enter both User ID and password.";
  }

  const request: { [key: string]: any } = {
    new_password: user.password, // Do not reset password
    phone_number: "1234567890",
    device_os: Device.osName,
    os_version: Device.osVersion,
    product: Device.modelName,
    brand: Device.brand,
    manufacturer: Device.manufacturer,
    model: Device.modelId,
    beiwe_version: "1.6.0",
  };

  try {
    let endpoint = "/register_user";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    await makePostRequestAsync(endpoint, request, {}, user);
  } catch (e) {
    return `Request error: ${e}.`;
  }

  await storeUserAsync(user);
  return null;
}

export async function getAllDataAsync(): Promise<UploadData> {
  const user = await getUserAsync();
  const pings = await getPingsAsync();
  const answers = await getAnswersAsync();

  const data = {
    patientId: user!.patientId,
    pings,
    answers,
  };
  return data;
}

export async function uploadDataAsync() {
  const user = (await getUserAsync())!;
  const data = await getAllDataAsync();

  try {
    let endpoint = "/ssnl_upload";
    if (Platform.OS === "ios") {
      endpoint += "/ios/";
    }
    const response = await makePostRequestAsync(endpoint, {}, data, user);
    //console.warn(`YAYA ${JSON.stringify(response)}`);
    return { status: "success", response };
  } catch (e) {
    return { status: "error", error: `Request error: ${e}.` };
  }
}

function base64ToBase64URL(input: string): string {
  return input.replace(/\+/g, "-").replace(/\//g, "_");
}

export async function getRequestURLAsync(
  endpoint: string,
  request: { [key: string]: any } = {},
  forUser?: User,
): Promise<string> {
  let user: User | null = forUser || null;
  if (!user) {
    user = await getUserAsync();

    if (user == null) {
      throw new Error("user == null in getRequestURLAsync");
    }
  }

  const passwordHash = await Crypto.digestStringAsync(
    Crypto.CryptoDigestAlgorithm.SHA256,
    user.password,
    {
      encoding: Crypto.CryptoEncoding.BASE64,
    },
  );
  request["password"] = base64ToBase64URL(passwordHash);
  request["device_id"] = `${user.patientId}-${Constants.installationId}`;
  request["patient_id"] = user.patientId;

  //console.warn(`request is ${JSON.stringify(request)}`);

  const serverUrl = await getServerUrlAsync();
  const query = Object.keys(request)
    .map((k) => encodeURIComponent(k) + "=" + encodeURIComponent(request[k]))
    .join("&");
  const url = `${serverUrl}${endpoint}?${query}`;
  return url;
}

export async function makePostRequestAsync(
  endpoint: string,
  request: { [key: string]: any },
  body?: { [key: string]: any },
  user?: User,
): Promise<any> {
  const headers = {
    "Beiwe-Api-Version": "2",
    Accept: "application/vnd.beiwe.api.v2, application/json",
  };

  const url = await getRequestURLAsync(endpoint, request, user);

  //console.warn("url is " + url);

  const response = await fetch(url, {
    method: "POST",
    headers,
    body: JSON.stringify(body),
  });

  if (response.status < 200 || response.status >= 400) {
    console.warn(`ERRORREQUEST: ${JSON.stringify(response)}`);

    if (response.status === 401 || response.status === 403) {
      throw new Error(
        `Verification failed.\n\nRequest: ${JSON.stringify(
          request,
        )}.\n\nResponse: ${JSON.stringify(response)}.`,
      );
    }

    throw new Error(
      `Request not successful.\n\nRequest: ${JSON.stringify(
        request,
      )}.\n\nResponse: ${JSON.stringify(response)}.`,
    );
  }

  //console.warn(response);

  return response.json();
}
