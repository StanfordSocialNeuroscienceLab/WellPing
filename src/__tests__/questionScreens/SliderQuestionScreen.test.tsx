import React from "react";
import { render, fireEvent, A11yAPI } from "react-native-testing-library";
import { ReactTestInstance } from "react-test-renderer";

import { SliderAnswerData } from "../../helpers/answerTypes";
import { QuestionType } from "../../helpers/helpers";
import { SliderQuestion } from "../../helpers/types";
import SliderQuestionScreen, {
  DEFAULT_SLIDER_VALUE,
  getQuestionDefaultSliderValue,
} from "../../questionScreens/SliderQuestionScreen";

const SLIDER_QUESTIONS: {
  [questionId: string]: SliderQuestion;
} = {
  WithoutDefault: {
    id: "WithoutDefault",
    type: QuestionType.Slider,
    question: "This question does not have any default value settings.",
    slider: ["leftmost", "rightmost"],
    next: "WithConstantDefault",
  },
  WithConstantDefault: {
    id: "WithConstantDefault",
    type: QuestionType.Slider,
    question: "This question has a constant default value.",
    defaultValue: 25,
    slider: ["left", "right"],
    next: "WithDefaultFromElsewhere",
  },
  WithDefaultFromElsewhere: {
    id: "WithDefaultFromElsewhere",
    type: QuestionType.Slider,
    question:
      "This question has a default value that is from another question.",
    defaultValueFromQuestionId: "WithConstantDefault",
    defaultValue: 100, // This should be ignored for this question.
    slider: ["left", "right"],
    next: null,
  },
};
const WITHOUT_DEFAULT = SLIDER_QUESTIONS["WithoutDefault"];
const WITH_CONSTANT_DEFAULT = SLIDER_QUESTIONS["WithConstantDefault"];
const WITH_DEFAULT_FROM_ELSEWHERE =
  SLIDER_QUESTIONS["WithDefaultFromElsewhere"];

const SLIDER_A11Y_LABEL = "slider input";
const getSlider = (getAllByA11yLabel: A11yAPI["getAllByA11yLabel"]) => {
  const sliderInputs = getAllByA11yLabel(SLIDER_A11Y_LABEL);
  expect(sliderInputs).toHaveLength(1);
  const sliderInput = sliderInputs[0];
  return sliderInput;
};
const moveSlider = (sliderInput: ReactTestInstance, value: number) => {
  fireEvent(sliderInput, "onSlidingComplete", {
    nativeEvent: { value },
  });
};

const simplePipeInExtraMetaData = (id: string) => id;

test("without default value", () => {
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const { getAllByA11yLabel, toJSON } = render(
    <SliderQuestionScreen
      key={WITHOUT_DEFAULT.id}
      question={WITHOUT_DEFAULT}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );

  // Because there isn't a need to pipe in any data.
  expect(mockPipeInExtraMetaData).not.toHaveBeenCalled();

  // Because there isn't any requirement to validate the data.
  expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

  const sliderInput = getSlider(getAllByA11yLabel);
  expect(sliderInput.props.value).toBe(DEFAULT_SLIDER_VALUE);

  expect(toJSON()).toMatchSnapshot();
});

test("with constant default value", () => {
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const { getAllByA11yLabel, toJSON } = render(
    <SliderQuestionScreen
      key={WITH_CONSTANT_DEFAULT.id}
      question={WITH_CONSTANT_DEFAULT}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );

  // Because there isn't a need to pipe in any data.
  expect(mockPipeInExtraMetaData).not.toHaveBeenCalled();

  // Because there isn't any requirement to validate the data.
  expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

  const sliderInput = getSlider(getAllByA11yLabel);
  expect(sliderInput.props.value).toBe(WITH_CONSTANT_DEFAULT.defaultValue);

  expect(toJSON()).toMatchSnapshot();
});

const testDefaultValueFromElsewhere = (
  prevAnswerData: SliderAnswerData | null,
) => {
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const { getAllByA11yLabel, toJSON } = render(
    <SliderQuestionScreen
      key={WITH_DEFAULT_FROM_ELSEWHERE.id}
      question={WITH_DEFAULT_FROM_ELSEWHERE}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{
        // @ts-ignore (we don't need to craft an entire AnswerEntity)
        WithConstantDefault: {
          data: prevAnswerData,
        },
      }}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );

  // To pipe in `defaultValueFromQuestionId`.
  expect(mockPipeInExtraMetaData).toHaveBeenCalledTimes(1);

  // Because there isn't any requirement to validate the data.
  expect(mockSetDataValidationFunction).not.toHaveBeenCalled();

  const sliderInput = getSlider(getAllByA11yLabel);
  // `defaultValue` should be ignored here.
  expect(sliderInput.props.value).not.toBe(
    WITH_DEFAULT_FROM_ELSEWHERE.defaultValue,
  );

  expect(toJSON()).toMatchSnapshot();

  return sliderInput;
};

test("with default value from elsewhere - with previous data", () => {
  const PREV_ANSWER_DATA = 33;
  const sliderInput = testDefaultValueFromElsewhere({
    value: PREV_ANSWER_DATA,
  });
  expect(sliderInput.props.value).toBe(PREV_ANSWER_DATA);
});

test("with default value from elsewhere - without previous data", () => {
  const sliderInput = testDefaultValueFromElsewhere(null);
  expect(sliderInput.props.value).toBe(
    // If previous data is `null`, the question should use the default
    // slider value of previous question.
    getQuestionDefaultSliderValue(
      SLIDER_QUESTIONS[WITH_DEFAULT_FROM_ELSEWHERE.defaultValueFromQuestionId!],
    ),
  );
});

test("update values", () => {
  const mockOnDataChangeFn = jest.fn();
  const mockPipeInExtraMetaData = jest.fn(simplePipeInExtraMetaData);
  const mockSetDataValidationFunction = jest.fn();

  const { getAllByA11yLabel } = render(
    <SliderQuestionScreen
      key={WITHOUT_DEFAULT.id}
      question={WITHOUT_DEFAULT}
      onDataChange={mockOnDataChangeFn}
      allAnswers={{}}
      allQuestions={SLIDER_QUESTIONS}
      pipeInExtraMetaData={mockPipeInExtraMetaData}
      setDataValidationFunction={mockSetDataValidationFunction}
    />,
  );

  const sliderInput = getSlider(getAllByA11yLabel);

  const NEW_VALUES = [88, 99, 66, 55];
  for (let i = 0; i < NEW_VALUES.length; i++) {
    const NEW_VALUE = NEW_VALUES[i];
    moveSlider(sliderInput, NEW_VALUE);
    expect(mockOnDataChangeFn).toHaveBeenNthCalledWith(i + 1, {
      value: NEW_VALUE,
    } as SliderAnswerData);
    expect(mockOnDataChangeFn).toHaveBeenCalledTimes(i + 1);
  }
});
