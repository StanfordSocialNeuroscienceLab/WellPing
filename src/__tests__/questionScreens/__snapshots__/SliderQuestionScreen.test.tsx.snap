// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with constant default value 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RCTSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={null}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={null}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    value={25}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question with default value with data \`{ value: 61 }\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RCTSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={null}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={null}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    value={61}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question with default value with data \`null\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RCTSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={null}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={null}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    value={100}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question without default value with data \`{ value: 31 }\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RCTSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={null}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={null}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    value={31}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`with default value from a question without default value with data \`null\` 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RCTSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={null}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={null}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    value={50}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      left
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      right
    </Text>
  </View>
</View>
`;

exports[`without default value 1`] = `
<View
  style={
    Object {
      "paddingVertical": 30,
    }
  }
>
  <RCTSlider
    accessibilityLabel="slider input"
    disabled={false}
    enabled={true}
    maximumTrackTintColor="#2F2424"
    maximumValue={100}
    minimumTrackTintColor="#2F2424"
    minimumValue={0}
    onChange={null}
    onResponderTerminationRequest={[Function]}
    onSlidingComplete={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={null}
    step={1}
    style={
      Object {
        "height": 40,
      }
    }
    value={50}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      leftmost
    </Text>
    <Text
      style={
        Object {
          "maxWidth": "40%",
        }
      }
    >
      rightmost
    </Text>
  </View>
</View>
`;
